### PR TITLE
questions in cfg file

### PR DESCRIPTION
**Pull request by baferpro**

### CODE REVIEW

**Comment by Bara** [comment]
Why russian text https://github.com/shanapu/MyJailbreak/pull/283/files#diff-e36baa1be9b148e27364c00347d02ae0 ?

File: addons/sourcemod/scripting/MyJailbreak/ratio.sp [hunk 1]
```diff
@@ -481,7 +476,7 @@ public Action Command_JoinGuardQueue(int client, int iArgNum)
 
 		if (iIndex == -1)
 		{
-			if ((gc_iJoinMode.IntValue == 0) || (gc_bAdminBypass.BoolValue && CheckVipFlag(client, g_sAdminFlag)))
+			if ((gc_iJoinMode.IntValue == 0) || (gc_bAdminBypass.BoolValue && CheckVipFlag(client, g_sAdminFlag))
```
Comment (Bara) [inline addons/sourcemod/scripting/MyJailbreak/ratio.sp]: Why you remove this ) ?

File: addons/sourcemod/scripting/MyJailbreak/ratio.sp [hunk 2]
```diff
@@ -635,7 +630,7 @@ public Action Command_ToggleRatio(int client, int args)
 		else
 		{
 			g_bRatioEnable = true;
-			CPrintToChatAll("%s %t", g_sPrefix, "ratio_hasactivated");
+			CPrintToChatAll("%s% t", g_sPrefix, "ratio_hasactivated");
```
Comment (Bara) [inline addons/sourcemod/scripting/MyJailbreak/ratio.sp]: "%s% t" should be "%s %t

File: addons/sourcemod/scripting/MyJailbreak/ratio.sp [hunk 3]
```diff
@@ -717,7 +712,7 @@ public Action Event_OnFullConnect(Event event, const char[] name, bool dontBroad
 	if (!gc_bForceTConnect.BoolValue || !g_bRatioEnable)
 		return Plugin_Continue;
 
-	if (!gc_bAdminBypass.BoolValue || !CheckVipFlag(client, g_sAdminFlag))
+	if (!gc_bAdminBypass.BoolValue || !CheckVipFlag(client, g_sAdminFlag)
```
Comment (Bara) [inline addons/sourcemod/scripting/MyJailbreak/ratio.sp]: Readd ) here too

File: addons/sourcemod/scripting/MyJailbreak/ratio.sp [hunk 4]
```diff
@@ -781,7 +776,7 @@ public Action Event_OnJoinTeam(int client, const char[] szCommand, int iArgCount
 
 		if (iIndex == -1)
 		{
-			if ((gc_iJoinMode.IntValue == 0) || (gc_bAdminBypass.BoolValue && CheckVipFlag(client, g_sAdminFlag)))
+			if ((gc_iJoinMode.IntValue == 0) || (gc_bAdminBypass.BoolValue && CheckVipFlag(client, g_sAdminFlag))
```
Comment (Bara) [inline addons/sourcemod/scripting/MyJailbreak/ratio.sp]: Readd ) here too

File: addons/sourcemod/scripting/MyJailbreak/ratio.sp [hunk 5]
```diff
@@ -809,7 +804,7 @@ public Action Event_OnJoinTeam(int client, const char[] szCommand, int iArgCount
 		return Plugin_Handled;
 	}
 
-	if ((gc_iJoinMode.IntValue == 0) || (gc_bAdminBypass.BoolValue && CheckVipFlag(client, g_sAdminFlag)))
+	if ((gc_iJoinMode.IntValue == 0) || (gc_bAdminBypass.BoolValue && CheckVipFlag(client, g_sAdminFlag))
```
Comment (Bara) [inline addons/sourcemod/scripting/MyJailbreak/ratio.sp]: Readd ) here too

File: addons/sourcemod/scripting/MyJailbreak/ratio.sp [hunk 6]
```diff
@@ -963,58 +958,129 @@ public int Handler_AcceptGuardRules(Handle menu, MenuAction action, int param1,
 
 void Menu_GuardQuestions(int client)
 {
-	char info[64], random[64];
+	char szPath[256];
+	BuildPath(Path_SM, szPath, sizeof(szPath), "configs/MyJailbreak/questions.ini");
+	KeyValues hKeyValues = new KeyValues("GlobalKey");
```
Comment (Bara) [inline addons/sourcemod/scripting/MyJailbreak/ratio.sp]: You should close hKeyValues to prevent memory leaks

**Comment by Bara** [comment]
and you've errors during compiling: https://travis-ci.org/shanapu/MyJailbreak/jobs/348993313

File: addons/sourcemod/scripting/MyJailbreak/ratio.sp [hunk 7]
```diff
@@ -971,16 +971,15 @@ void Menu_GuardQuestions(int client)
...
 	Panel InfoPanel = new Panel();
 	char szRandomQuestion[16];
 	IntToString(GetRandomInt(1, i), szRandomQuestion, sizeof(szRandomQuestion));
 	g_iRandomAnswer[client] = GetRandomInt(1, 3);
 	hKeyValues.Rewind();
 	if(hKeyValues.JumpToKey(szRandomQuestion))
 	{
-		char info[128];
-		char szBuffer[256];
+		char ;
```
Comment (Bara) [inline addons/sourcemod/scripting/MyJailbreak/ratio.sp]: Remove this line (error 001: expected token: "-identifier-", but found ";")

File: addons/sourcemod/scripting/MyJailbreak/ratio.sp [hunk 8]
```diff
@@ -1026,6 +1026,11 @@ void Menu_GuardQuestions(int client)
 
 		InfoPanel.Send(client, Handler_GuardQuestions, 20);
 	}
+	
+	if (hKeyValues)
```
Comment (Bara) [inline addons/sourcemod/scripting/MyJailbreak/ratio.sp]: You don't need to check if handle is valid. That's just required with CloseHandle and was "removed" with delete.

Comment (baferpro) [reply addons/sourcemod/scripting/MyJailbreak/ratio.sp]: just ```delete hKeyValues;```?

File: addons/sourcemod/scripting/MyJailbreak/ratio.sp [hunk 9]
```diff
@@ -147,7 +147,7 @@ public void OnPluginStart()
 	gc_bToggleAnnounce = AutoExecConfig_CreateConVar("sm_ratio_disable_announce", "0", "Announce in a chatmessage on roundend when ratio is disabled", _, true, 0.0, true, 1.0);
 	gc_bAdsVIP = AutoExecConfig_CreateConVar("sm_ratio_adsvip", "1", "0 - disabled, 1 - enable adverstiment for 'VIPs moved to front of queue' when player types !guard ", _, true, 0.0, true, 1.0);
 	gc_iJoinMode = AutoExecConfig_CreateConVar("sm_ratio_join_mode", "1", "0 - instandly join ct/queue, no confirmation / 1 - confirm rules / 2 - Qualification questions", _, true, 0.0, true, 2.0);
-	gc_iQuestionTimes = AutoExecConfig_CreateConVar("sm_ratio_questions", "3", "How many question a player have to answer before join ct/queue. need sm_ratio_join_mode 2", _, true, 1.0, true, 5.0);
+	gc_iQuestionTimes = AutoExecConfig_CreateConVar("sm_ratio_questions", "20", "How many question a player have to answer before join ct/queue. need sm_ratio_join_mode 2", _, true, 1.0, true, 20.0);
```
Comment (Hexer10) [inline addons/sourcemod/scripting/MyJailbreak/ratio.sp]: What? 20 questions?

File: addons/sourcemod/scripting/MyJailbreak/ratio.sp [hunk 10]
```diff
@@ -181,26 +181,21 @@ public void OnSettingChanged(Handle convar, const char[] oldValue, const char[]
 	{
 		strcopy(g_sAdminFlag, sizeof(g_sAdminFlag), newValue);
 	}
-	else if (convar == gc_sAdminFlag)
-	{
-		strcopy(g_sAdminFlag, sizeof(g_sAdminFlag), newValue);
-	}
```
Comment (Hexer10) [inline addons/sourcemod/scripting/MyJailbreak/ratio.sp]: You might leave here as it is now

File: addons/sourcemod/scripting/MyJailbreak/ratio.sp [hunk 11]
```diff
@@ -536,7 +531,7 @@ public Action Command_JoinGuardQueue(int client, int iArgNum)
 						SetArrayCell(g_aGuardQueue, 0, client);
 					}
 
-					CReplyToCommand(client, "%s %t", g_sPrefix, "ratio_thxvip");
+					CReplyToCommand(client, "%s %t", g_sPrefix, "ratio_thxadmin");
```
Comment (Hexer10) [inline addons/sourcemod/scripting/MyJailbreak/ratio.sp]: "ratio_thxadmin" doesn't even exists in the phrases file

File: addons/sourcemod/scripting/MyJailbreak/ratio.sp [hunk 12]
```diff
@@ -963,58 +958,132 @@ public int Handler_AcceptGuardRules(Handle menu, MenuAction action, int param1,
 
 void Menu_GuardQuestions(int client)
 {
-	char info[64], random[64];
+	char szPath[256];
+	BuildPath(Path_SM, szPath, sizeof(szPath), "configs/MyJailbreak/questions.ini");
+	KeyValues hKeyValues = new KeyValues("GlobalKey");
+	hKeyValues.ImportFromFile(szPath);
+	int i = 0;
+	hKeyValues.Rewind();
```
Comment (Hexer10) [inline addons/sourcemod/scripting/MyJailbreak/ratio.sp]: Afaik you don't need this here

File: addons/sourcemod/scripting/MyJailbreak/ratio.sp (hunk 8 above)
Comment (Bara) [reply addons/sourcemod/scripting/MyJailbreak/ratio.sp]: no just the if check, keep delete.

Comment (baferpro) [reply addons/sourcemod/scripting/MyJailbreak/ratio.sp]: https://github.com/shanapu/MyJailbreak/pull/283/commits/e89d76230afe2bb9a5692d71337e88ffeeec5f54
Same?

File: addons/sourcemod/scripting/MyJailbreak/ratio.sp (hunk 12 above)
Comment (baferpro) [reply addons/sourcemod/scripting/MyJailbreak/ratio.sp]: What exactly?

Comment (Hexer10) [reply addons/sourcemod/scripting/MyJailbreak/ratio.sp]: hKeyValues.Rewind();

File: addons/sourcemod/scripting/MyJailbreak/ratio.sp [hunk 13]
```diff
@@ -147,7 +147,7 @@ public void OnPluginStart()
 	gc_bToggleAnnounce = AutoExecConfig_CreateConVar("sm_ratio_disable_announce", "0", "Announce in a chatmessage on roundend when ratio is disabled", _, true, 0.0, true, 1.0);
 	gc_bAdsVIP = AutoExecConfig_CreateConVar("sm_ratio_adsvip", "1", "0 - disabled, 1 - enable adverstiment for 'VIPs moved to front of queue' when player types !guard ", _, true, 0.0, true, 1.0);
 	gc_iJoinMode = AutoExecConfig_CreateConVar("sm_ratio_join_mode", "1", "0 - instandly join ct/queue, no confirmation / 1 - confirm rules / 2 - Qualification questions", _, true, 0.0, true, 2.0);
-	gc_iQuestionTimes = AutoExecConfig_CreateConVar("sm_ratio_questions", "3", "How many question a player have to answer before join ct/queue. need sm_ratio_join_mode 2", _, true, 1.0, true, 5.0);
+	gc_iQuestionTimes = AutoExecConfig_CreateConVar("sm_ratio_questions", "3", "How many question a player have to answer before join ct/queue. need sm_ratio_join_mode 2", _, true, 1.0, true, 20.0);
```
Comment (shanapu) [inline addons/sourcemod/scripting/MyJailbreak/ratio.sp]: we should remove the max value (or increase it). maybe some servers want to use more than 20.

Comment (Bara) [reply addons/sourcemod/scripting/MyJailbreak/ratio.sp]: Or just remove the max. value (no limit)

**Comment by shanapu** [comment]
Great Thanks bafer for your PR.
Also great thanks to bara and hexah for the first support.

I think I would be better to cache all the questions & answers instead of reading the kv.file every time again. 

Otherwise, your work looks very good. I will happily merge it when the questions are cached.
Thanks

File: addons/sourcemod/scripting/MyJailbreak/ratio.sp [hunk 14]
```diff
@@ -90,7 +90,7 @@ Handle hKeyValues;
 // Integer
 int g_iRandomAnswer[MAXPLAYERS+1];
 int g_iQuestionTimes[MAXPLAYERS+1];
-int countQuestions = 0;
+int countQuestions;
```
Comment (shanapu) [inline addons/sourcemod/scripting/MyJailbreak/ratio.sp]: please name it **g_i**CountQuestions or **g_i**QuestionsCount

'g_' stand for global variable
'i' stand for a integer varible

**Comment by shanapu** [comment]
Sry I hadn't the time for a full review with test last weeks.
When I have done it I will merge the PR even when you don't have cached the questions.

I assume you didn't know what I mean or how to solve it, when I said it should be cached.
To be honest I also don't know 'the best way' from out of my head to cache the questions & answers.

An [Array](https://sm.alliedmods.net/new-api/adt_array) / [Trie](https://sm.alliedmods.net/new-api/adt_trie) or even something else...?!

Or maybe it is unnecessary to cache it. right now, IDK.

**Comment by shanapu** [comment]
Thanks for all your work guys! It's merged. 
With time I(we) will include a client cookie so players don't have to answer question every time when solved already. 
And I have to think about the caching of questions.

Thanks you!

**Pull request merged**